feat(Forms): add `transformData` to the onSubmit event listener of Form.Handler (#4183)
dnbexperience/eufemia: packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/Form/Handler/info.mdx

- `filterData` Filters the given/internal data set.
- `reduceToVisibleFields` Reduces the given data set to only contain the visible fields (mounted fields).

> - `transformData` Will call the given function for each data path. The returned value will be used as the new value.

Reviewer: ```suggestion
- `transformData` Will call the given function for each data path. Can be used to...(something more here)...
```

Given the sentence `The returned value will be used as the new value.`, I feel like one could mistakenly think that it somehow changes/updates the `value`, which I don't think happens, hence why I try to change this text.

  )
}
```

#### `transformData`

The `transformData` function will call the given function for each data path. The returned value will be used as the new value.

The callback function receives the following arguments as an object:

- `path` The path of the field.
- `value` The value of the field.
- `displayValue` The displayed value of the field.
- `label` The label of the field.
- `props` The given field properties.
- `internal` The `internal` parameter contains `{ error: Error | undefined }` you can utilize if needed.

**displayValue** will be `undefined` for most fields.

- It will return the displayed title for [Selection](/uilib/extensions/forms/base-fields/Selection/) fields as a string (or React.ReactNode).

> - It will return the displayed content for [ArraySelection](/uilib/extensions/forms/base-fields/ArraySelection/) fields as an array that contains a string (or React.ReactNode).

Reviewer: Great explanation for this and Selection. Do we have Unit tests that confirms/prove the same? Could be nice to have.

Author: Yes, we test it in the field tests. But we could add one for the array value, in the provider test file as well 👍

Reviewer: Should we also add support for `displayValue` in the following components?
- [Field.Toggle](https://eufemia.dnb.no/uilib/extensions/forms/base-fields/Toggle/demos/) when using `textOn`/`falseText ` and not having a `label`?
- [Field.Boolean](https://eufemia.dnb.no/uilib/extensions/forms/base-fields/Boolean/demos/) when using `trueText `/`textOff` and not having a `label`?
- [Field.SelectCountry](https://eufemia.dnb.no/uilib/extensions/forms/feature-fields/SelectCountry/)
- [Field.PhoneNumber](https://eufemia.dnb.no/uilib/extensions/forms/feature-fields/PhoneNumber/demos/)? 
- Also, will this works with all variants of [Field.Selection](https://eufemia.dnb.no/uilib/extensions/forms/base-fields/Selection/properties/)?

  )
}
```

#### `transformData`

The `transformData` function will call the given function for each data path. The returned `value` will replace each data entry. It's up to you to define the shape of the value.

The callback function receives the following arguments as an object:

- `path` The path of the field.
- `value` The value of the field.
- `displayValue` The displayed value of the field.
- `label` The label of the field.
- `props` The given field properties.
- `internal` The `internal` parameter contains `{ error: Error | undefined }` you can utilize if needed.

> If a field does not provide a **displayValue**, the `value` will be returned instead.

Author: True. Thinking of it again, I think we should return "undefined". It's more "workable" and predictable that way.

> **displayValue** can be `undefined` if a filed does not support it, or it's value is not set (`emptyValue`).

Reviewer: ```suggestion
**displayValue** can be `undefined` if a field does not support it, or it's value is not set (`emptyValue`).
```

How can I know if a given field supports displayValue or not?

> Most of the fields will return the displayed value as a string. But there are some exceptions:

Reviewer: ```suggestion
Most of the fields will return the `displayValue` as a string. But there are some exceptions:
```

  )
}
```

#### `transformData`

> The `transformData` function will call the given function for each data path. The returned value will be used as the new value.

Reviewer: Update here as well, if we change here: https://github.com/dnbexperience/eufemia/pull/4183/files#r1820372758

The callback function receives the following arguments as an object:

- `path` The path of the field.
- `value` The value of the field.
- `displayValue` The displayed value of the field.
- `label` The label of the field.
- `props` The given field properties.
- `internal` The `internal` parameter contains `{ error: Error | undefined }` you can utilize if needed.

> **displayValue** will be `undefined` for most fields.

Reviewer: Will it be `undefined` for most fields now after the latest changes? 🤔 
Now that we've included it in so many fields?

Reviewer: Since we've now added it to `Field.String` all the fields that use Field.String(StringField local name in our code) in our code, will return a `displayValue`:
- `BankAccountNumber`
- `Email`
- `Name`
- `NationalIdentityNumber`
- `OrganizationNumber`
- `Password`
- `PhoneNumber`
- `PostalCodeAndCity`

Does it make sense to list which fields that supports `displayValue` now? Isn't it all fields?

> **displayValue** can be `undefined` if a filed does not support it, or it's value is not set (`emptyValue`).

Reviewer: Since we've now added it to `Field.String` all the fields that use Field.String(StringField local name in our code) in our code, will return a `displayValue`:
- `BankAccountNumber`
- `Email`
- `Name`
- `NationalIdentityNumber`
- `OrganizationNumber`
- `Password`
- `PhoneNumber`
- `PostalCodeAndCity`

Author: Yes

Author: I'v updated the description:

> Most of the fields will return the displayed value as a string. But there are some exceptions:
> - It will return the displayed/active options content for [ArraySelection](/uilib/extensions/forms/base-fields/ArraySelection/) fields as an array that contains a string (or React.ReactNode).

- It will return the displayed/active option content for [Selection](/uilib/extensions/forms/base-fields/Selection/) fields as a string (or React.ReactNode).
- It will return the displayed/active options content for [ArraySelection](/uilib/extensions/forms/base-fields/ArraySelection/) fields as an array that contains a string (or React.ReactNode).
- It will return the displayed text value for [Toggle](/uilib/extensions/forms/base-fields/Toggle/) field as a string (or React.ReactNode).
- It will return the displayed text value for [Boolean](/uilib/extensions/forms/base-fields/Toggle/) field as a string.
- It will return the displayed value (including the mask) for [String](/uilib/extensions/forms/base-fields/String/) field as a string.
- It will return the displayed value (including the mask) for [Number](/uilib/extensions/forms/base-fields/Number/) field as a string.
- It will return the displayed value (including the mask) for [Currency](/uilib/extensions/forms/feature-fields/Currency/) field as a string.
- It will return the displayed value (including the mask) for [PhoneNumber](/uilib/extensions/forms/feature-fields/PhoneNumber/) field as a string.

> - It will return the selected country name for [SelectCountry](/uilib/extensions/forms/feature-fields/SelectCountry/) field as a string.

Reviewer: In regards to the [comment](https://github.com/dnbexperience/eufemia/pull/4183/files#r1820440374), and this list, does it make sense to list which fields that supports `displayValue` now? Isn't it all fields? 🤔

Author: Yes, very correct. But I don't want to have it listed, because it will quickly get out of sync. Any thoughts on such an argument?

Reviewer: Yeah, that will quickly get out of sync.
Could all fields support it?
Or doesn't that make sense?

Initially, I was against setting `displayValue` to fallback to be `value` for the fields that do not "have a `displayValue`".
But now it feels like we do?

Take the following [scenario](https://codesandbox.io/p/sandbox/throbbing-water-5sfg4k?workspaceId=9eb653cb-4406-4b8f-bc96-714927cc7fcd), using Field.String:
`<Field.String label="Field.String" value="myValue" path="/myString" />`

Both `value` and `displayValue` is the same.

Should we just set `displayValue` to fallback to be `value` for the Fields that do not "have or need `displayValue`"? And in the docs state that `displayValue` is the "human value", and should only be used in these scenarios, for the real value use `value`(this is better tested, documented, etc, etc).

Reviewer: Use case for Field.Slider where there's a need for displayValue(since it can be different from the value):
- When using `numberFormat={{ currency: 'EUR' }}`

Reviewer: Use case for `Field.Date` where there's a need for `displayValue`(since it can be different from the `value`):
- In the following [example](https://eufemia.dnb.no/uilib/extensions/forms/feature-fields/Date/#label-and-value), where `value` would be `2023-01-16,` and `displayValue` would be the formatted date `16/01/2023`.

Reviewer: Use case for `Field.Expiry` where there's a need for `displayValue`(since it can be different from the `value`):
- In the following [example](https://eufemia.dnb.no/uilib/extensions/forms/feature-fields/Expiry/demos/#label), where `value` would be `0835,` and `displayValue` would be the formatted date `08/35`.

Reviewer: If we want to add support for `displayValue` to the fields above, we could also do it progressively(over time), and not only in this PR.

Author: Thank you so much for the good cases. I added all of them ✅

Author: > But now it feels like we actually do fallback to setting displayValue to be value for the fields that do not "have a displayValue" 🤔

I added commit where we always fall back to `value`, when `displayValue` is not given.
